File: Clean-Architecture/Feature/Javascript/example/src/features/auth/domain/usecases/LoginUser.js
import { AppError } from '../../../../core/errors/AppError.js';

export class LoginUser {
  constructor({ authRepository }) {
    this.authRepository = authRepository;
  }

  async execute({ email, password }) {
    const normalizedEmail = email.trim().toLowerCase();
    const result = await this.authRepository.login({
      email: normalizedEmail,
      password,
    });

    if (!result) {
      throw new AppError('Invalid email or password', {
        statusCode: 401,
        code: 'INVALID_CREDENTIALS',
      });
    }

    return result;
  }
}

File: Clean-Architecture/Feature/dart/example/lib/features/auth/domain/usecases/login_user.dart
import '../../../../core/result/result.dart';
import '../entities/auth_session.dart';
import '../repositories/auth_repository.dart';

class LoginUser {
  const LoginUser(this._repository);
  final AuthRepository _repository;

  Future<Result<AuthSession>> call({required String email, required String password}) {
    return _repository.login(
      email: email.trim().toLowerCase(),
      password: password,
    );
  }
}
